stablehlo/dialect/VhloAttrs.td

```diff
@@ -204,4 +168,99 @@ def VHLO_ConvolutionAttributes {
   );
 }
 
+//===----------------------------------------------------------------------===//
+// Forked Attributes
+//===----------------------------------------------------------------------===//
+
+def VHLO_ArrayAttrV1 : VHLO_AttrDef<"ArrayV1"> {
+  let mnemonic = "array";
+  let parameters = (ins ArrayRefParameter<"mlir::Attribute">:$value);
+  let genVerifyDecl = 1;
+  let extraClassDefinition = [{
+    LogicalResult ArrayV1Attr::verify(
+        llvm::function_ref<mlir::InFlightDiagnostic ()> errFn, ArrayRef<mlir::Attribute> value) {
+      if (!allFromVhlo(value)) return errFn() << "expected array of VHLO attriutes";
+      return success();
+    }
+  }];
+  let assemblyFormat = "`<` custom<AttributeArray>($value) `>`";
+}
+
+def VHLO_DenseIntOrFPElementsData : AttrParameter<"::llvm::ArrayRef<char>", "Array of int"> {
+  // Custom allocator to copy dense elements data into MLIR Context
+  let allocator = "$_dst = $_allocator.copyInto($_self);";
+}
+def VHLO_DenseIntOrFPElementsAttrV1 : VHLO_AttrDef<"DenseIntOrFPElementsV1"> {
+  let mnemonic = "dense";
+  let parameters = (ins "::mlir::Type":$type, VHLO_DenseIntOrFPElementsData:$raw_data);
+  let genVerifyDecl = 1;
+  let extraClassDefinition = [{
+    LogicalResult DenseIntOrFPElementsV1Attr::verify(
+        llvm::function_ref<mlir::InFlightDiagnostic ()> errFn, mlir::Type type, ArrayRef<char>) {
+      if (!isFromVhlo(type)) errFn() << "expected VHLO type";
+      return success();
+    }
+  }];
+  let hasCustomAssemblyFormat = 1;
+}
+
+def VHLO_FlatSymbolRefAttrV1 : VHLO_AttrDef<"FlatSymbolRefV1"> {
+  let mnemonic = "sym";
+  let parameters = (ins "::mlir::Attribute":$root_reference);
+  let genVerifyDecl = 1;
+  let extraClassDefinition = [{
+    LogicalResult FlatSymbolRefV1Attr::verify(
+        llvm::function_ref<mlir::InFlightDiagnostic ()> errFn, mlir::Attribute rootReference) {
+      if (!isFromVhlo(rootReference)) return errFn() << "expected VHLO attribute";
+      return success();
+    }
+  }];
+  let assemblyFormat = "`<` $root_reference `>`";
+}
+
+def VHLO_FloatAttrV1 : VHLO_AttrDef<"FloatV1"> {
+  let mnemonic = "float";
+  let parameters = (ins "mlir::Type":$type, VHLO_DoubleV1:$value);
+  let genVerifyDecl = 1;
+  let extraClassDefinition = [{
+    LogicalResult FloatV1Attr::verify(
+        llvm::function_ref<mlir::InFlightDiagnostic ()> errFn, mlir::Type type, APFloat value) {
+      if (!isFromVhlo(type)) return errFn() << "expected VHLO float type";
```

Comment:
Sounds good. Another issue I'm having is that these `errFn` produce multiple line error messages, and I'm unable to get `-verify-diagnostics` to be happy with multi-line messages. These all work, but verifiers are untested currently. I have a separate file with test cases.

Reply:
Updated the error message as suggested.

Per my follow-up question, these test points were deleted in this PR but added back as legalization tests (not verifier tests) in https://github.com/openxla/stablehlo/pull/925/files#diff-a0dbcff346cd4a84a6267d153cdf699aa68d7672270b80ec0a7aadb21debe099

This was done because that PR adds FP8, so we are able to test that `0.3.0` fails on downgrade when that type is used.